[#22] 사용자/친구 차단 기능 구현

msugo1 opened the pull request:
## 📖  Controller 계층 관련

### 📑 FriendController

-  차단 요청을 전달 받을 `blockUser` 메소드 추가

## 📖 Service 계층 관련

### 📑 FriendService

- 차단 요청에 대한 비즈니스 로직을 처리할 `blockUser` 메소드 추가

- 로직이 `중복`되는 부분(자기 자신인지 확인, 존재하는 사용자인지 확인)을 `별도의 메소드`로 분리

## 📖 Mapper 계층 관련

### 📑 friendMapper

- 차단 처리 내용을 DB에 반영할 `blockUser` 메소드 추가

### 📑 friendMapper.xml

- 위의 내용에 대한 쿼리 추가

## 📖 예외 관련

### 📑 예외 이름 변경

- 친구 추가 요청뿐만 아니라 차단 시에도 함께 사용할 수 있도록 `Duplicate/Invalid FriendRequestException -> Duplicate/Invalid
  RequestException`으로 이름 변경

- 같은 용도로 쓰이는 여러 개의 예외를 `별도의 패키지`로 이동

## 📖 테스트 관련

- 차단 로직을 처리하는 메소드에 대한 테스트 작성

- 포스트맨을 사용해 전체 테스트

Commit:
[#22] 사용자 차단 기능 구현

Controller 계층 관련

* FriendController

- 차단 요청을 전달 받을 `blockUser` 메소드 추가

Service 계층 관련

* FriendService

- 차단 요청에 대한 비즈니스 로직을 처리할 `blockUser` 메소드 추가

- 로직이 `중복`되는 부분(자기 자신인지 확인, 존재하는 사용자인지 확인)을
  `별도의 메소드`로 분리

Mapper 계층 관련

* friendMapper

- 차단 처리 내용을 DB에 반영할 `blockUser` 메소드 추가

* friendMapper.xml

- 위의 내용에 대한 쿼리 추가

예외 관련

* 예외 이름 변경

- 친구 추가 요청뿐만 아니라 차단 시에도 함께 사용할 수 있도록
  Duplicate/Invalid FriendRequestException을`Duplicate/Invalid
  RequestException`으로 이름 변경

- 한 용도로 쓰이는 여러 개의 예외를 별도의 패키지로 이동

테스트 관련

- 차단 로직을 처리하는 메소드에 대한 테스트 작성

- 포스트맨을 사용해 전체 테스트

Commit:
[#23] 부하분산을 위해 마스터-슬레이브 구조로 DB변경

Configuration

* ReplicationRoutingDataSource

- AbstractRoutingDataSource 를 상속받고 읽기 전용 트랜잭션에 여부에 따라 `동적으로 master/slave 분기를 나눠주는` 메소드인 `determineCurrentLookupKey()`를 구현한 클래스

* DataSourceConfig

- 설정된 Master/Slave Datasource를 각 `분기에 맞게` 나눠주는`routingDataSource` 메소드 추가

- lazy Loading과Routing Data Source 를 올바르게 동작할 수 있도록
  도와주는 Proxy Data Source 를 생성할 `proxyDataSource` 메소드 추가

- 위의 DataSource를 활용해 트랜잭션 매니저 빈을 생성할
  `transactionManager` 메소드 추가

- DataSourceConfig 내에 존재하는 메소드들이 거의 모두 DataSource를
  리턴하기 때문에 빈 속성이 겹치는 문제가 발생하여 `name`을 활용해
  고유한 이름을 부여하고 `Qualifier`를 활용해 맞는 이름을 가진 빈이
  주입될 수 있도록 수정

Service 계층 관련

- 데이터 정합성이 엄격하지 않은 데이터를 읽어오는 메소드에
  `@Transactional(readOnly=true)`를 부여해 슬레이브에서 데이터를 읽어올
  수 있도록 설정

* 포스트맨을 활용해 작동 테스트

msugo1 commented:
### 📝질문
- 사용자 차단을 위한 Service를 따로 두려고 구상을 해봤는데요. 문제점이 두 가지가 있었습니다. 
1. `FriendService`와 요구되는 메소드가 겹치는 `중복되는 부분`이 많고, Block 사용자를 위한 테이블을 따로 만들면 `friends` 테이블에 있는 `status`와도 내용이 겹치고 차단한 사용자를 확인해야 하니 DB와의 `커넥션 한 번 더 추가`가 필요한 문제

2. FriendService에서 차단 관련 테이블을 검색하기 위해서 `차단 관련 서비스를 주입`받고 반대로 차단 관련 서비스에서 차단 시 친구 삭제를 위해 `FriendService 주입`받는 경우 `순환참조`의 문제

- 그래서 FriendService 내에 차단관련 기능을 두는 것이 최선인 것 같아서 그렇게 구현하였는데요. 성격이 다른 것 같으면서도 완전히 다르지도 않은 것 같아서 그런데 차단 관련 서비스를 따로 만들어서 분리를 하는 것이 좋을까요? 아니면 이대로 가도 될까요?

f-lab-dev commented:
차단은 친구뿐 아니라 친구가 아닌 유저들도 차단이 가능합니다~ 묶는거보단 분리하는게 낫지 않을까요?

Commit:
[#22] 피드백 반영 - 사용자 차단 기능 분리

Controller 계층 관련

* BlockUserController 클래스 생성

- 사용자 차단 요청을 전달 받을 `blockUser` 메소드 추가

* FriendController

- 사용자 차단 요청을 전담하던 메소드 삭제

Enum 클래스 관련

* FriendStatus

- 차단 상태를 나타내던 `BLOCK` 상수 삭제

- 상수 이름 변경

Mapper 계층 관련

* FriendMapper

- 차단 요청을 DB에 반영하던 `blockUser` 메소드 삭제

* BlockUserMapper 인터페이스 생성

- 사용자가 이미 차단되었는지 확인하기 위한 `isUserBlocked` 메소드 추가

- 사용자 차단 로직 처리 내용을 DB에 반영할 `blockUser` 메소드 추가

* BlockUserMapper.xml

- 위의 로직을 처리할 쿼리문장 추가

Service 계층 관련

* FriendService

- 사용자 차단과 관련된 로직 삭제

- `getStatus` 메소드를 통해 친구 상태 중 차단을 의미하는 BLOCK
  리턴여부를 확인하는 로직을 제거

- 대신 `blockUserService`를 주입 받아서 사용자 차단 여부를 확인하도록
  변경

- blockUserServic와 FriendService에서 모두 사용되는 코드를 하나로
  통합하기 위해서 별도의 클래스로 분리(자기 자신에 대한 요청인지, 중복된
  요청인지 확인하는 부분)

- `validateStatusDetail` 메소드에서 switch  상태 값 확인 부분에 있던 `BLOCK` 삭제

* BlockUserService

- 사용자가 이미 차단된 상태인지 확인하는 로직을 담당하는 `isUserBlocked`
  메소드 추가

- 사용자 차단 로직을 처리할 `blockUser` 추가

- 순환참조의 문제 때문에 friendService를 주입받지 못하고
  `friendMapper`로 대체해서 주입

예외 관련

* InvalidRequestException, DuplicateRequestException

- 범용적으로 사용할 수 있도록 메소드 이름 자체를 해당 예외 이름으로
  바꾸고 예외를 핸들링할 시 리턴하는 제목도 범용적인 제목으로 변경
  (세부적인 내용은 여전히 예외를 던지는 메소드에서 전달할 수 있으므로)

Util 관련

* TargetValidator 클래스 추가

- 타겟 사용자가 `자기 자신`인지 혹은 `실제로 존재`하는지를 확인하기
  위한 중복되는 메소드를 하나로 모을 클래스

테스트 관련

* FriendServiceTest 변경

- 내부 로직 수정으로 인해 테스트 내용 변경

- 사용자 차단과 관련된 테스트 부분 삭제

* BlockUserServiceTest 추가

- 사용자 차단과 관련된 테스트 작성

* 전체적인 로직을 포스트맨을 사용해 테스트

Commit:
[#21] 업데이트 반영위한 merge(/20 -> /21)

Commit:
Merge branch 'feature/21' into feature/22

Commit:
Merge branch 'feature/22' into feature/23

Commit:
[#23] 피드백 반영

- 하드 코딩되어 있는 DB 종류 `master`, `slave`를 상수처리

msugo1 commented:
> 차단은 친구뿐 아니라 친구가 아닌 유저들도 차단이 가능합니다~ 묶는거보단 분리하는게 낫지 않을까요?

일단 분리는 해봤는데 여기저기 구조를 변경해가면서 구상해보고 시도해봤는데 순환참조 구조를 해결할 수가 없어서 `friendMapper` 를 `blockUserService` 에서 불러오는 것으로 합의를 봐야했습니다. 혹시 좀 더 좋은 방법이 있으면 조언해주시면 감사하겠습니다.

Commit:
[#24] 사용자 차단 해제 기능 구현

Controller 계층 관련

* BlockUserController

- 사용자 차단 해제 요청을 받을 `unblockUser` 메소드 추가

Service 계층 관련

* BlockUserService

- 사용자 차단 해제에 요청에 대한 비즈니스 로직을 처리 할 `unblockUser`
  메소드 추가

Mapper 계층 관련

* BlockUserMapper

- DB에 차단 해제 내역을 반영 할`deleteBlockUser` 메소드 추가

- DB 관련 행동임을 더 잘 나타내기 위해서 `blockUser ->
  insertBlockUser`로 메소드 이름 변경

테스트 관련

- 친구 차단 해제에 관련된 테스트 작성

- 포스트맨을 활용해 전체적인 테스트

Commit:
[#24] 가독성을 높이기 위해 쿼리문 줄 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/model/friend/FriendListRequest.java:
```diff
@@ -21,12 +20,12 @@
                                 .userId(userId)
-                                .offset(MAX_PER_PAGE_FRIEND * (Math.max(pageNumber, 1) - 1))
-                                .limit(MAX_PER_PAGE_FRIEND)
+                                .offset(pagination.getMaxPageFriend() * (Math.max(pageNumber, 1) - 1))
```
이런 페이징 계산식 같은 경우는 따로 클래스로 분리해보면 좋을 것 같네요~

msugo1 replied:
최대 페이지 값만 담고 있던 `pagination` 클래스에 `caculateOffset` 메소드를 추가해서 페이징 계산식도 처리할 수 있게끔 변경하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FriendService.java:
```diff
@@ -42,57 +46,56 @@ public FriendStatus getFriendStatus(String userId, String targetId) {
+                throw new DuplicateRequestException("이미 해당 사용자로부터 친구추가 요청을 받은 상태입니다. 받은 친구 요청을 다시 확인해주세요.");
+            case FRIEND:
+                throw new DuplicateRequestException("이미 친구로 등록된 사용자에게 다시 친구 요청을 보낼 수 없습니다.");
+            case NONE:
+            default:
+                break;
```
이 부분은 생략이 가능해보입니다~

msugo1 replied:
저 부분이 없어도 스위치 문을 통과하는 것은 처음 알게 되었네요! 감사합니다. 피드백 반영했습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/Pagination.java:
```diff
@@ -1,6 +1,16 @@
+import org.springframework.stereotype.Component;
+
+@Getter
+@Component
```
아마 `maxPageFriend` 때문에 bean으로 정의하신 것 같은데 값만 담는 용도로는 굳이 이렇게 클래스를 분리하지 않으셔도 됩니다~

msugo1 replied:
빈으로 등록한게 해당 클래스 인스턴스를 한 곳에만 여러 곳에 쓰니까 매번 새로운 객체를 만드는 것보다  싱글턴 빈으로 등록해서 주입받으면 메모리 관리에도 더 효율적이겠다는 생각에 빈으로 만들었거든요. 읽기 전용 클래스이니 멀티 스레드 환경에서도 문제가 될 일도 없을 거 같기도 하구요.

페이징 계산식도 해당 클래스에 추가했는데, 빈으로 놔두는 것이 좋을까요? 아니면 그래도 빈보다는 일반 클래스로 분류하는 것이 좋을까요?

msugo1 committed:
Merge pull request #52 from f-lab-edu/feature/23

[#23] 부하분산을 위해 마스터-슬레이브 구조로 DB변경

msugo1 committed:
Merge pull request #54 from f-lab-edu/feature/24

[#24] 사용자 차단 해제 기능 구현

Commit:
[#22] 피드백 반영

1. 페이징 계산식을 별도로 분리

- 페이징을 담당하는 클래스인 pagination에 `caculateOffset` 추가

2. 스위치 문 내의 불필요한 조건 생략

Commit:
Merge 'feature/22' into origin/feature/22 for push
